Add image for hive3 and hadoop3

stagraqubole opened the pull request:
Images based on hdp3.

I have use image created via this change for Hive ACID tests in https://github.com/prestosql/presto/pull/1034

findepi commented on prestodev/centos7-oj8/README.md:
```diff
@@ -0,0 +1,13 @@
+
+## Oracle license
+
+By using this image, you accept the Oracle Binary Code License Agreement for Java SE available here:
```
Doesn't apply, the image uses openjdk

findepi commented on prestodev/centos7-oj8/README.md:
```diff
@@ -0,0 +1,13 @@
+# centos7-oj8 [![][layers-badge]][layers-link] [![][version-badge]][dockerhub-link]
+
+[layers-badge]: https://images.microbadger.com/badges/image/prestodev/centos7-oj8.svg
```
@kokosing i'd like to abandon this READMEs. I don't find them useful.
WDYT?

stagraqubole replied:
I noticed the link are dead as of now but wasnt sure if there are plan to get them working so extended the readmes from existing ones.

findepi replied:
@stagraqubole i understand. You did the reasonable thing. Let's wait for @kokosing 's opinion before dropping this.

kokosing replied:
I agree. Especially, if they have outdated information then they do more harm than good.

stagraqubole replied:
I will remove these then.

findepi commented on prestodev/centos7-oj8/Dockerfile:
```diff
@@ -0,0 +1,40 @@
+    sed -i 's/https/http/g' /etc/yum.repos.d/epel.repo && \
+    yum install -y python-pip && \
+    pip install --upgrade pip==9.0.3 `# latest version for Python 2.6` && \
+    pip install --upgrade setuptools==36.8.0 `# latest version for Python 2.6` && \
```
> latest version for Python 2.6

I hope Centos 7 has python 2.7, does it?
If so, version pinning is no longer needed (at least not to these specific versions)

stagraqubole replied:
I will check that

stagraqubole replied:
These are the defaults present or yum installed in centos7:
python: 2.7.5
pip: 8.1.2
setuptools: 0.9.8

I removed these two lines for specific versions and supervisord installs and works fine without them.

findepi commented on prestodev/centos7-oj8/Dockerfile:
```diff
@@ -0,0 +1,40 @@
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+FROM library/centos:7
```
Just curious -- can hdp3 be run on centos6?

stagraqubole replied:
I am forgetting the details, but I wasnt able to get it working on centos6

findepi commented on prestodev/centos7-oj8/Dockerfile:
```diff
@@ -0,0 +1,40 @@
+    \
+    # install supervisor
+    yum --enablerepo=extras install -y setuptools epel-release && \
+    sed -i 's/https/http/g' /etc/yum.repos.d/epel.repo && \
```
Should not be needed. Why so?

stagraqubole replied:
will double check.

stagraqubole replied:
> Thanks @stagraqubole for working on this!
> 
> By any chance, did you also prepare kerberized images?
> They would be required when we want to run all the tests on hdp3.

I didnt prepare kerberized image, my requirement was limited to running tests using hive3.

findepi commented:
Thanks @stagraqubole for working on this!

By any chance, did you also prepare kerberized images?
They would be required when we want to run all the tests on hdp3.

stagraqubole commented:
> My understanding is that most of the files were created as a copy from hdp2.6 and then applying certain changes. Now, these changes are not visible.
> 
> Could you please make 2 separate commits? One that is plain copy and second which changes the file contents as necessary?

Makes sense, I will separate it out.

stagraqubole commented:
Updates as per the comments @findepi

kokosing commented on prestodev/hdp3-hive3/Dockerfile:
```diff
@@ -10,17 +10,23 @@
 
 # Install Hadoop, Hive (w/ MySQL)
 RUN yum install -y \
+    yum-plugin-ovl \
```
what is this?

stagraqubole replied:
I will check again, might not be needed anymore, had done multiple rounds of this to get things right and this might be an artifact of one of those runs.

kokosing commented on prestodev/hdp3-hive3/Dockerfile:
```diff
@@ -59,7 +66,7 @@ RUN chmod 755 /root && chmod 700 /root/.ssh
 
 # HDFS ports
-EXPOSE 1004 1006 8020 50010 50020 50070 50075 50470
+EXPOSE 1004 1006 8020 9866 9867 9870 9864 50470
```
What happened to these ports? Is 50070 no longer exposed? Was it migrated somehow? I used to use it in other versions.

stagraqubole replied:
Port have changed. I used https://issues.apache.org/jira/browse/HDFS-9427 and updated the ports that were in use.

kokosing commented on prestodev/hdp3-hive3/files/etc/hadoop/conf/yarn-site.xml:
```diff
@@ -87,4 +87,15 @@
+    </property>
+
+    <property>
+        <name>yarn.scheduler.minimum-allocation-mb </name>
```
extra space

kokosing commented on prestodev/hdp3-hive3/files/etc/hadoop/conf/yarn-site.xml:
```diff
@@ -87,4 +87,15 @@
+
+    <property>
+        <name>yarn.scheduler.minimum-allocation-mb </name>
+        <value>256</value>
```
no description here? or remove description from the above entry.

kokosing commented on prestodev/hdp3-hive3/files/etc/tez/conf/tez-site.xml:
```diff
@@ -0,0 +1,147 @@
+
+<!-- Put site-specific property overrides in this file. -->
+
+<configuration>
```
formatting here is off

kokosing commented on prestodev/hdp3-hive3/files/etc/tez/conf/tez-site.xml:
```diff
@@ -0,0 +1,147 @@
+  <name>tez.shuffle-vertex-manager.desired-task-input-size</name>
+  <value>true</value>
+</property>
+-->
```
remove commented code

kokosing commented on prestodev/hdp3-hive3/files/etc/tez/conf/tez-site.xml:
```diff
@@ -0,0 +1,147 @@
+  <name>tez.runtime.shuffle.buffersize</name>
+  <value>8192</value>
+</property>
+-->
```
same here

kokosing commented on prestodev/hdp3-hive3/files/etc/tez/conf/tez-site.xml:
```diff
@@ -0,0 +1,147 @@
+  <name>shuffle.memory.limit.percent</name>
+  <value>0.5</value>
+</property>
+-->
```
same here

stagraqubole commented:
Handled your comments @kokosing

kokosing commented:
Ah... I forgot about tests. Please see `https://github.com/prestosql/docker-images/blob/master/bin/test.sh`. It would be great if you could add a test for this image. Where you could do some simple queries against hive (CREATE TABLE, INSERT, SELECT, TRUNCATE, DROP).

stagraqubole commented:
> Ah... I forgot about tests. Please see `https://github.com/prestosql/docker-images/blob/master/bin/test.sh`. It would be great if you could add a test for this image. Where you could do some simple queries against hive (CREATE TABLE, INSERT, SELECT, TRUNCATE, DROP).

@kokosing I see that the test.sh file already has the test queries and can run for any profile which has compose file defined. Do you want me to make changes to ensure that it runs on hdp3-hive3 or something more?

stagraqubole commented:
- Added docker-compose.yml for hdp3-hive3 for tests
- Added beeline-site.xml to let `hive -e` work without specifying connection url

Ran `bin/test.sh hdp3-hive3` successfully. Let me know if something more is needed @kokosing

stagraqubole commented:
I also noticed that in test.sh, `set +e` is set around the function that runs the hive queries. Wouldnt we want the test run to fail if hive queries fail for some reason?

stagraqubole commented:
> I also noticed that in test.sh, `set +e` is set around the function that runs the hive queries. Wouldnt we want the test run to fail if hive queries fail for some reason?

The expectation is to return the error code and do cleanup, hence the `set +e`. But we are currently returning last hive query's error code which is not correct. I have added a commit to fix that as well

findepi commented on prestodev/hdp3-hive3/Dockerfile:
```diff
@@ -0,0 +1,73 @@
+# Licensed under the Apache License, Version 2.0 (the "License");
```
> Replicate hdp2.6-hive as the starting point for hdp3-hive3

- Claudera may release `hdp 3.2` which we may want to support separately, just as we support 2.5 and 2.6 separately
- HDP 3.1 includes Hive 3.1.0 only, so suffixing "hive" with "3" is not necessary

Rename `hdp3-hive3` to `hdp3.1-hive` (commit message, folder name)

stagraqubole replied:
Done

findepi commented on prestodev/centos7-oj8/Dockerfile:
```diff
@@ -0,0 +1,37 @@
+FROM library/centos:7
+MAINTAINER Presto community <https://prestosql.io/community.html>
+
+# Install Java and presto-admin dependences (already has python 2.6)
```
Remove `(already has python 2.6)`

stagraqubole replied:
Done

findepi commented on prestodev/centos7-oj8/Dockerfile:
```diff
@@ -0,0 +1,37 @@
+    # install supervisor
+    yum --enablerepo=extras install -y setuptools epel-release && \
+    yum install -y python-pip && \
+    pip install supervisor==3.4.0 && \
```
`supervisor` was pinned to this version as this was the last v compatible with Python 2.6.
Consider unpinning.

findepi commented on prestodev/hdp3-hive3/Dockerfile:
```diff
@@ -10,14 +10,19 @@
+# mysql
+RUN wget http://repo.mysql.com/mysql-community-release-el7-5.noarch.rpm && \
+  rpm -ivh mysql-community-release-el7-5.noarch.rpm && \
+  yum update -y
```
You can just 
```
yum localinstall http://repo.mysql.com/mysql-community-release-el7-5.noarch.rpm
```
(when you `rpm -i`, then `yum` complaints afterwards)


What's the reason for not using what's bundled with Centos? They include `mariadb-server` which is likely a drop-in replacement.

stagraqubole replied:
It is not available in centos7

findepi replied:
I tried `docker run --rm -it centos:7 yum install -y mariadb-server` and it worked. (Output a but too long to quote it here 😉 )

stagraqubole replied:
My comment was about mysql-server not being bundled with Centos :)

Why prefer mariadb over mysql? I feel mysql is more common and hence easier to fix issues around it during container setup and later for debugging if needed.

findepi replied:
Because it's easy to install and doesn't require downloading .rpm over http.

> easier to fix issues around it during container setup

if you run into issues, let's reconsider. Does it sound OK?

stagraqubole replied:
Ok, i will give it a shot.

stagraqubole replied:
You were right, mariadb worked out smoothly. It also helped remove the permission hack in mysql supervisord config. I have made the change to use mariadb-server.

findepi commented on prestodev/hdp3-hive3/files/etc/hadoop/conf/yarn-site.xml:
```diff
@@ -87,4 +87,16 @@
+    <property>
+        <name>yarn.scheduler.minimum-allocation-mb</name>
+        <value>256</value>
+        <description>Minimum container memory, default value is 1024 which is too high for the test queries</description>
```
This is not the original description, it's your commentary.
Move commentary to `<!-- .... -->` comment.
Otherwise a reader may overlook you added some useful information here.

findepi commented on prestodev/hdp3-hive3/files/etc/hive/conf/hive-site.xml:
```diff
@@ -64,4 +64,34 @@
         <value>hdfs,hive</value>
     </property>
 
+    <!--  https://stackoverflow.com/questions/47287359/hive-reach-max-worker-and-cannot-connect-to-hiveserver2 -->
```
Move the comment before `<value>NOSASL</value>` line, so that it's clear it applies to this configuration  only and not some following.

findepi commented on prestodev/hdp3-hive3/files/etc/hive/conf/hive-site.xml:
```diff
@@ -64,4 +64,34 @@
+    <!--  https://stackoverflow.com/questions/47287359/hive-reach-max-worker-and-cannot-connect-to-hiveserver2 -->
+    <property>
+      <name>hive.server2.authentication</name>
+      <value>NOSASL</value>
```
Why not `NONE`? (i checked the SO link but it's unclear to me)

stagraqubole replied:
NONE worked, I guess the client I was working with originally didnt work with NOSASL. Moved to NONE now.

findepi commented on prestodev/hdp3-hive3/files/etc/hive/conf/hive-site.xml:
```diff
@@ -64,4 +64,34 @@
+
+     <property>
+        <name>hive.server2.enable.doAs</name>
+        <value>false</value>
```
Isn't this default?

stagraqubole replied:
Default is true but to make things work with NOSASL this needed to be overridden. 
I will check the NOSASL vs NONE part, if we can use NONE then this part can go away.

stagraqubole replied:
Removed

findepi commented on prestodev/hdp3-hive3/files/etc/hive/conf/hive-site.xml:
```diff
@@ -64,4 +64,34 @@
+
+    <property>
+      <name>hive.support.concurrency</name>
+      <value>true</value>
```
What is this for?
(I think we'll never run any concurrent workload in hive directly.)

stagraqubole replied:
This is a requirement for using DbTxnManager and DbTxnManager is a requirement for ACID.

findepi commented on prestodev/hdp3-hive3/files/etc/supervisord.d/mysql-metastore.conf:
```diff
@@ -1,5 +1,5 @@
 [program:mysql-metastore]
-command=/usr/bin/pidproxy /var/run/mysqld/mysqld.pid /usr/bin/mysqld_safe
+command=/bin/bash -c "(chown -R mysql:mysql /var/lib/mysql) && (/usr/bin/pidproxy /var/run/mysqld/mysqld.pid /usr/bin/mysqld_safe)"
```
I guess this is not sufficient to `chown` once in Dockerfile. Did you try? i wouldn't be surprised -- i think i encountered such a problem recently somewhere else.

no need for `( ... )`, please remove them

stagraqubole replied:
I wasnt able to get it running with `chown` in Dockerfile. Tried again by placing another instance of it after full setup but still there are failures when container starts.

stagraqubole commented:
@findepi handled your latest comments. Only the part about mysql vs mariadb is left.

stagraqubole commented:
@findepi @kokosing ping!! I have a PR in presto blocked on this, can we move forward with this please.

kokosing commented:
> @findepi @kokosing ping!! I have a PR in presto blocked on this, can we move forward with this please.

Can you please post a docker image under different repository name to docker hub for now? I would like to merge this PR and the one in Presto together. Notice that merging this PR is not enough, we still need to do a release.

stagraqubole commented:
> > @findepi @kokosing ping!! I have a PR in presto blocked on this, can we move forward with this please.
> 
> Can you please post a docker image under different repository name to docker hub for now? I would like to merge this PR and the one in Presto together. Notice that merging this PR is not enough, we still need to do a release.

I have done that, PR is open in presto (with a bit outdated docker image): https://github.com/prestosql/presto/pull/1034

I will update it with the latest image.

kokosing commented:
> I will update it with the latest image.

Super. Thanks!

stagraqubole commented:
> > I will update it with the latest image.
> 
> Super. Thanks!

This is done.

findepi commented:
I manually tested locally the metastore part of the image

```
docker run --rm -it --name hadoop-master --hostname hadoop-master -p 9083:9083 -p 1180:1180 prestodev/hdp3.1-hive:latest
```

```
echo "127.0.0.1 hadoop-master" > /etc/hosts
```

```
connector.name=hive-hadoop2
hive.metastore.uri=thrift://hadoop-master:9083
hive.metastore.thrift.client.socks-proxy=hadoop-master:1180
```

It worked like a charm.

Good job!

findepi commented:
i will take over this PR & merge it, if you don't mind.

findepi commented on prestodev/hdp3.1-hive/files/root/setup.sh:
```diff
@@ -39,15 +45,21 @@ killall java
+chown -R mysql:mysql /var/lib/mysql
+chgrp -R mysql /var/lib/mysql
+rm -f  /var/lib/mysql/ib_logfile0
+rm -f  /var/lib/mysql/ib_logfile1
```
i removed  these `rm`s

findepi commented on prestodev/hdp3.1-hive/files/etc/tez/conf/tez-site.xml:
```diff
@@ -0,0 +1,104 @@
+    </property>
+    <property>
+        <name>tez.task.resource.memory.mb</name>
+        <value>1024</value>
```
made it 128

findepi commented on prestodev/hdp3.1-hive/files/etc/tez/conf/tez-site.xml:
```diff
@@ -0,0 +1,104 @@
+    </property>
+    <property>
+        <name>tez.am.resource.memory.mb</name>
+        <value>1024</value>
```
made it 128

findepi commented on prestodev/hdp3.1-hive/files/etc/tez/conf/tez-site.xml:
```diff
@@ -0,0 +1,104 @@
+    </property>
+    <property>
+        <name>tez.runtime.io.sort.mb</name>
+        <value>512</value>
```
made it 8

findepi commented:
Thanks @stagraqubole, this is merged as https://github.com/prestosql/docker-images/commit/51c48e9c5cc9090382ad1f203e45a5042a55c8c2
